benchmarks/instruction_counts/execution/work.py

```diff
@@ -0,0 +1,204 @@
+
+        cmd.extend([
+            _PYTHON, WORKER_PATH,
+            "--communication_file", self._communication_file,
+        ])
+        return " ".join(cmd)
+
+    @property
+    def duration(self) -> float:
+        return (self._end_time or time.time()) - self._start_time
+
+    @property
+    def result(self) -> Union[WorkerOutput, WorkerFailure]:
+        self._maybe_collect()
+        assert self._result is not None
+        return self._result
+
+    def poll(self) -> Optional[int]:
+        self._maybe_collect()
+        return self._retcode
+
+    def interrupt(self) -> None:
+        """Soft interrupt. Allows subprocess to cleanup."""
+        self._proc.send_signal(signal.SIGINT)
+
+    def terminate(self) -> None:
+        """Hard interrupt. Immediately SIGTERM subprocess."""
+        self._proc.terminate()
+
+    def _maybe_collect(self) -> None:
+        if self._result is not None:
+            # We've already collected the results.
+            return
+
+        self._retcode = self._proc.poll()
+        if self._retcode is None:
+            # `_proc` is still running
+            return
+
+        with open(self._communication_file, "rb") as f:
+            result = WorkerUnpickler(f).load_output()
+
+        if isinstance(result, WorkerOutput) and self._retcode:
+            # Worker managed to complete the designated task, but worker
+            # process did not finish cleanly.
+            result = WorkerFailure("Worker failed silently.")
+
+        if isinstance(result, WorkerTimerArgs):
+            # Worker failed, but did not write a result so we're left with the
+            # original TimerArgs. Grabbing all of stdout and stderr isn't
+            # ideal, but we don't have a better way to determine what to keep.
+            proc_stdout = self._proc.stdout
+            assert proc_stdout is not None
+            result = WorkerFailure(
+                failure_trace=proc_stdout.read().decode("utf-8"))
+
+        self._result = result
+        self._end_time = time.time()
+
+        # Release communication file.
+        os.remove(self._communication_file)
+
+
+class InProgress:
+    """Used by the benchmark runner to track outstanding jobs.
+    This class handles bookkeeping and timeout + retry logic.
+    """
+    _proc: _BenchmarkProcess
+    _timeouts: int = 0
+
+    def __init__(self, work_order: WorkOrder, cpu_list: Optional[str]):
+        self._work_order = work_order
+        self._proc = _BenchmarkProcess(work_order, cpu_list)
+
+    @property
+    def work_order(self) -> WorkOrder:
+        return self._proc._work_order
+
+    @property
+    def cpu_list(self) -> Optional[str]:
+        return self._proc._cpu_list
+
+    @property
+    def proc(self) -> _BenchmarkProcess:
+        # NB: For cleanup only.
+        return self._proc
+
+    @property
+    def duration(self) -> float:
+        return self._proc.duration
+
+    @property
+    def ready(self) -> bool:
```

Comment:
I would suggest this is not a property -- properties are idempotent - this has semantics which vary based on when / how many times it is called. Also, the semantics are kind of confusing here because it combines as a check with control logic. What options we have for structuring this?

Reply:
I would argue that a property is only idempotent if the underlying structure has `const` semantics (whatever that means in Python...), but I suppose that's a question for the philosophers. I changed the name to `check_finished` and removed the property decorator to signal that it's a more active call. Thoughts?